README version 0.1 (#15)
Joffref/genz: README.md

# GenZ

[![Go Reference](https://pkg.go.dev/badge/github.com/Joffref/genz.svg)](https://pkg.go.dev/github.com/Joffref/genz)
[![Go Report Card](https://goreportcard.com/badge/github.com/Joffref/genz)](https://goreportcard.com/report/github.com/Joffref/genz)
[![License](https://img.shields.io/github/license/Joffref/genz)](LICENSE)
![GitHub stars](https://img.shields.io/github/stars/Joffref/genz)

> At this moment, this project is in development. It is not ready for production use. See the [Contributing](#contributing) section if you would like to help make it better.

✨ Never write toil/boilerplate code again! ✨

> Thanks to GenZ, you can generate code for your own needs, without having to write a single line of code.

Reviewer: Technically... our templates are a superset of Go code, and there will be Go tests, but it's okay !

<p align="center" heigth="50%">
    <img heigth="50%" src="./docs/assets/logo_500_500.png"> 
</p>

GenZ is a template-based generator for Go. A single binary that can be called with the native `go generate` to automate generation of your Go code based on templates.

## Installation

You can install GenZ in different ways, depending on your needs.
We strongly recommend using the Go version for production use.

### Using Go - Production

```bash
go install github.com/Joffref/genz
```

### From source - Development

```bash
git clone https://github.com/Joffref/genz.git
cd genz
make install
```

## Getting Started

Let's say you have several types defined in your project (such as `Human` as defined below).
```go
package main

type Human struct {
	// +required
	Firstname string
	// +required
	Lastname string
	Age uint
}
```
And you want to add a `Validate()` method on these type.

With GenZ, all you have to do is to write a template once:
```gotemplate
package main

import (
    "fmt"
)

func (v {{ .Type }}) Validate() error {
{{ range .Attributes }} {{$attribute := .}}
{{ if eq .Type "string" }}{{ range .Comments }}

> {{ if eq . "+required" }}

Reviewer: Example won't work, comment is equal to " +required" (with a space) in the example above

# GenZ
## Requirements

- [Go >1.20](https://go.dev/doc/install)
## Try it out

> Explore built-in `examples`, clone repo, and run

Reviewer: I would have kept this (invite the user to browse our examples) after the ## Getting started example

> At this moment, this project is in development. It is not ready for production use.

> > Refers to the [Roadmap](#roadmap) section for more information on what has been done.

Reviewer: We have no roadmap for now, I would remove it for now

>
> See the [Contributing](#contributing) section if you would like to help make it better.

<p align="center" width="100%">
    <img width="50%" src="./docs/assets/logo_500_500.png"> 
</p>

> GenZ is your all-in-one opinionated code generator for Go. It leverages the power of Go templates to generate code from your own templates.

Reviewer: ```suggestion
:sparkles: Never write toil/boilerplate code again! :sparkles: 
GenZ is a template-based generator for Go. A single binary that can be called with the native `go generate` to automate generation of Go tested code based on your templates.
```

(I would have put the Getting Started just after this, and installation at the end)

Thanks to GenZ, you can generate code for your own needs, without having to write a single line of code.

## Installation

You can install GenZ in different ways, depending on your needs.
We strongly recommend using the Go version for production use.

### Using Go - Production

```bash
go install github.com/Joffref/genz
```

### From source - Development

```bash
git clone https://github.com/Joffref/genz.git
cd genz
make install
```

> ## Usage

Reviewer: ```suggestion
## Getting Started
```

> Here is a simple example of how to use GenZ.
> 
> ### Go

Reviewer: Useless for me
```suggestion
```

> Let say you have a `main.tmpl` file in your project, with the following content:

Reviewer: The real usecase is not having a template and wanting to use it, this is the opposite.

I would start from "let's say you have several types defined in your project (such as `Human` as defined below)"
```go
package main

type Human struct {
	// +required
	Firstname string
	// +required
	Lastname string
	Age uint
}
```
And you want to add a `Validate()` method on those types.
With GenZ, all you have to do is to write a template once:

`validator.tmpl`
```gotemplate
package main
import (
    "fmt"
)
func (v {{ .Type }}) Validate() error {
{{ range .Attributes }} {{$attribute := .}}
{{ if eq .Type "string" }}{{ range .Comments }}
{{ if eq . "+required" }}
    if v.{{ $attribute.Name }} == "" {
        return fmt.Errorf("attribute '{{ $attribute.Name }}' must be set")
    }
{{ end }}{{ end }}{{ end }}{{ end }}
    return nil
}
```

And generate the associated Go code tied to your type :

1. Add `//go:generate genz -type Human -template validator.tmpl -output human.gen.go` on your go code.
2. Call `go generate ./...`

And your validator is generated :sparkles: 
```go
package main

import (
    "fmt"
)

func (v Human) Validate() error {
    if v.Firstname == "" {
        return fmt.Errorf("attribute 'Firstname' must be set")
    }
    if v.Lastname == "" {
        return fmt.Errorf("attribute 'Lastname' must be set")
    }
    return nil
}
```

Checkout other examples in [/examples](/examples)

Reviewer: Maybe a Getter is a simpler Getting started example don't you think ?
